internal/graph/singleflight_resolver.go

```diff
@@ -0,0 +1,118 @@
+package graph
+
+import (
+	"context"
+	"sync/atomic"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.uber.org/zap"
+
+	"resenje.org/singleflight"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/openfga/openfga/internal/build"
+	"github.com/openfga/openfga/pkg/logger"
+	"github.com/openfga/openfga/pkg/telemetry"
+)
+
+var (
+	deduplicatedDispatchCount = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: build.ProjectName,
+		Name:      "deduplicated_dispatch_count",
+		Help:      "The total number of calls to ResolveCheck that were prevented due to deduplication of singleflight resolver.",
+	})
+
+	deduplicatedDBQueryCount = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: build.ProjectName,
+		Name:      "deduplicated_datastore_query_count",
+		Help:      "The total number of datastore queries that were prevented due to deduplication of singleflight resolver.",
+	})
+)
+
+type SingleflightCheckResolver struct {
+	delegate CheckResolver
+	group    singleflight.Group[string, ResolveCheckResponse]
+	logger   logger.Logger
+}
+
+func NewSingleflightCheckResolver() *SingleflightCheckResolver {
+	s := &SingleflightCheckResolver{}
+	s.delegate = s
+
+	return s
+}
+
+// SetDelegate sets this SingleflightCheckResolver's dispatch delegate.
+func (s *SingleflightCheckResolver) SetDelegate(delegate CheckResolver) {
+	s.delegate = delegate
+}
+
+// Close implements CheckResolver.
+func (s *SingleflightCheckResolver) Close() {}
+
+// ResolveCheck implements CheckResolver.
+func (s *SingleflightCheckResolver) ResolveCheck(
+	ctx context.Context,
+	req *ResolveCheckRequest,
+) (*ResolveCheckResponse, error) {
+	ctx, span := tracer.Start(ctx, "ResolveCheck")
+	defer span.End()
+	span.SetAttributes(attribute.String("resolver_type", "SingleflightCheckResolver"))
+	span.SetAttributes(attribute.String("tuple_key", req.GetTupleKey().String()))
+
+	key, err := CheckRequestCacheKey(req)
+	if err != nil {
+		s.logger.Error("singleflight cache key computation failed with error", zap.Error(err))
```

Comment:
Given the log message will already contain the error I'd recommend changing the message every so slightly. 
```suggestion
		s.logger.Error("singleflight cache key computation failed", zap.Error(err))
```